CWE-316 regression: CommandScope.execute() finally-block clearCredentialArguments() breaks any caller re-using a CommandScope

## TL;DR

PR #7741 (CWE-316: clear password memory after use in legacy CLI + CommandScope, commit `4850342ae7`) added a `finally { clearCredentialArguments(); }` block at the end of `CommandScope.execute()`. That block overwrites every credential-bearing argument value (`password`, `passwd`, `secret`, `token`, `apikey`, `accesskey`, `licensekey`) with `"*****"` in the scope's `argumentValues` map.

This breaks **any caller that calls `execute()` more than once on the same `CommandScope`** — the second execute() reads `"*****"` instead of the real credential and fails to authenticate.

The canary test in this very PR — `liquibase.dbtest.pgsql.PostgreSQLIntegrationTest.testStatusRunDuringUpdate` — does exactly that (three `commandScope.execute()` calls in a row on the same scope) but apparently isn't run against a live postgres in the project's PR CI, so the regression slipped through.

## Repro / canary

`liquibase-integration-tests/src/test/java/liquibase/dbtest/pgsql/PostgreSQLIntegrationTest.java::testStatusRunDuringUpdate`:

```java
CommandScope commandScope = getStatusCommandScope(this.sleepChangelog);
CommandResults statusCheckBefore = commandScope.execute();   // PASSES; clearCredentialArguments() then wipes scope's password to "*****"

CompletableFuture<Boolean> updateFuture =
        CompletableFuture.supplyAsync(() -> { ...liquibase.update(...); });
Thread.sleep(3000);
CommandResults statusCheckDuring = commandScope.execute();   // FAILS — password is now "*****"
```

The second `execute()` triggers connection setup with the masked password and the JDBC driver rejects it:

```
FATAL: password authentication failed for user "lbuser"
at liquibase.dbtest.pgsql.PostgreSQLIntegrationTest.testStatusRunDuringUpdate(PostgreSQLIntegrationTest.java:96)
```

Pro's daily subtree-sync builds reproduce this consistently (4 out of 4 runs since 2026-05-22 fail this exact test, all others pass).

## Diagnosis

The fix in PR #7741 conflates two things:

1. *"Drop references so the credential String becomes GC-eligible after the scope is no longer in use."* — Useful for embedded long-running scenarios.
2. *"Drop references after every `execute()`."* — Not safe; nothing in the contract of `execute()` says the scope is single-use.

Code paths that legitimately reuse a scope:
- The test above (three sequential `execute()` calls on the same status-check scope).
- Any embedded integration that builds a scope once and triggers it on a schedule.
- Any flow that retries `execute()` after a transient error.

## Suggested fixes (pick one)

**Option A — make the clearing opt-in.** Add `clearCredentialArguments()` as a public method on `CommandScope`; remove the `finally` block. Callers (e.g. `Main.java`) call it explicitly at the end of their CLI session. Tests / long-lived integrations that reuse the scope don't suffer. Preserves the CWE-316 intent (CLI sessions still clear at the right boundary).

**Option B — gate the clearing on a flag.** Add `CommandScope.markSingleUse()`; only clear if that's been called. Backwards-compatible default.

**Option C — clear on scope close rather than per-execute.** If `CommandScope` had a close/dispose lifecycle, do the clearing there.

I'm happy to author a PR for Option A if that's the preferred direction — please let me know.

## Context

Discovered while sprint-defending the Pro-side subtree-sync of this commit batch into liquibase-pro. The five-PR security batch (#7741/#7743/#7744/#7747/#7748/#7750/#7756) is otherwise green — this is the only blocker.

## Correction: upstream `main` is currently red on this exact test — striking my earlier claim

In the issue body I wrote: *"upstream's main CI somehow doesn't run this test against a live postgres so the regression wasn't caught at PR-author time."* That is **not accurate** — apologies. Upstream main **does** run this test, and it **is** currently failing with the same signature. Posting the evidence so anyone reading this issue has the correct timeline.

### Direct evidence on `main`

`Unit and Integration Tests` workflow, push to `main`, commit `8ed866995e` (PR #7760 — the DAT-23096 fixture fix landing):

- Run: https://github.com/liquibase/liquibase/actions/runs/26402847894
- Failed job: `Integration Test (postgresql)` — https://github.com/liquibase/liquibase/actions/runs/26402847894/job/77720530913
- Failure:

  ```
  [ERROR] liquibase.dbtest.pgsql.PostgreSQLIntegrationTest.testStatusRunDuringUpdate -- Time elapsed: 8.249 s <<< ERROR!
  liquibase.exception.CommandExecutionException: ... FATAL: password authentication failed for user "lbuser"
      at liquibase.dbtest.pgsql.PostgreSQLIntegrationTest.lambda$testStatusRunDuringUpdate$0(PostgreSQLIntegrationTest.java:85)
      at liquibase.dbtest.pgsql.PostgreSQLIntegrationTest.testStatusRunDuringUpdate(PostgreSQLIntegrationTest.java:93)
  ```

Same test, same root cause as described in the issue body. The diagnosis (`finally { clearCredentialArguments(); }` wiping credentials on multi-execute scopes) stands.

### Why this regression was masked on `main` for four days

A second, unrelated breakage in the same window hid the integration-test failure until 2026-05-25 13:26 UTC:

- **2026-05-21 12:14-12:15 UTC** — CWE-78 #7747, CWE-470 #7748, CWE-22 #7750 landed three new `GlobalConfiguration` entries (`allowExecuteCommand`, `allowCustomChange`, `allowParentDirectoryReferences`) but none of them updated `LiquibaseCommandLineTest.expectedHelpOutput`. That made the `LiquibaseCommandLineTest.help output` UNIT test fail. The matrix `Run Test for (Java N <OS>)` jobs went red on every main commit from `26225284221` onward.
- The workflow's `Integration Test (postgresql)` job depends on the unit-test matrix succeeding, so it was **`skipped`** rather than `failure` on those runs. From a quick triage glance it looked like only unit tests were broken.
- **2026-05-25 13:26 UTC** — PR #7760 (DAT-23096) added the three missing stanzas; unit tests went green again; integration tests started running again; **and the CWE-316 regression became visible immediately** in the very next run on `main`.

So the regression has been *latent in main since* `4850342ae7` (2026-05-21 12:15 UTC, PR #7741) but *visible in main only since* `8ed866995e` (2026-05-25 13:26 UTC, PR #7760).

### One workflow-design observation while I was in there

PR #7741's pre-merge build doesn't include the `Unit and Integration Tests` workflow at all (the PR rollup only shows `claude-review` + a couple of unrelated checks). The workflow trigger appears to be `push` to `main`, not `pull_request`. That's why neither this regression nor the earlier unit-test fixture gap was visible to PR authors — they only fire after the merge button. Not asking you to redesign your CI here, just flagging that this gap is part of why three audit-batch PRs in a row landed without their snapshot test running, and why CWE-316 #7741 landed without anyone seeing it break the postgres integration test.

I'm still happy to author the upstream PR for Option A (or B / C if preferred) — let me know.

--- a/liquibase-standard/src/main/java/liquibase/command/CommandScope.java
+++ b/liquibase-standard/src/main/java/liquibase/command/CommandScope.java
@@ -52,9 +52,23 @@ public class CommandScope {
 
     /**
      * Substring tokens (lowercase) that mark an argument key as credential-bearing.
-     * Values for matching keys are overwritten with {@code "*****"} at the end of
-     * {@link #execute()} so the original credential Strings become GC-eligible and
-     * do not linger in heap for the rest of the JVM lifetime (CWE-316).
+     * Values for matching keys are overwritten with {@code "*****"} when the
+     * (public) {@link #clearCredentialArguments()} method is invoked by a caller.
+     * That makes the original credential Strings GC-eligible so they do not linger
+     * in heap for the rest of the JVM lifetime in long-running embedded scenarios
+     * (CWE-316).
+     * <p>
+     * <b>Caller-invoked, not auto-invoked from {@link #execute()}</b>. Calling it
+     * from an outer {@code finally} block in {@code execute()} (as the original
+     * CWE-316 fix did) breaks any caller that calls {@code execute()} more than
+     * once on the same {@code CommandScope} — the second call reads {@code "*****"}
+     * instead of the real credential and the JDBC driver rejects the connection.
+     * The CLI in {@code Main} wraps its single-use scopes in try-finally to call
+     * this method after each {@code execute()}; programmatic / library callers
+     * who reuse a {@code CommandScope} (e.g. status-before / update / status-after
+     * patterns) should call it once after the final {@code execute()}, never
+     * between executions. Discovered via
+     * {@code PostgreSQLIntegrationTest.testStatusRunDuringUpdate}.
      * <p>
      * Matching is intentionally substring-based on the lowercased key: keys like
      * {@code "argument__bearerToken"} are caught by {@code "token"};
@@ -379,22 +393,30 @@ public class CommandScope {
             throw e;
         } catch (Exception e) {
             throw new CommandExecutionException(e);
-        } finally {
-            clearCredentialArguments();
         }
     }
 
     /**
-     * Overwrite the values of any credential-bearing argument keys with
-     * {@code "*****"} so the original credential Strings stop being referenced by
-     * this scope. Strings are immutable and cannot be wiped in-place, but dropping
-     * the reference here makes them GC-eligible (assuming no other references hold
-     * the same String) instead of letting them live for the rest of the JVM in
-     * long-running embedded scenarios (CWE-316: Cleartext Storage of Sensitive
-     * Information in Memory). Called from {@link #execute()}'s outer finally so
-     * it runs regardless of command outcome; package-private for direct testing.
+     * Overwrite the values of any credential-bearing argument keys (per
+     * {@link #CREDENTIAL_KEY_TOKENS}) with {@code "*****"} so the original
+     * credential Strings stop being referenced by this scope. Strings are
+     * immutable and cannot be wiped in-place, but dropping the reference here
+     * makes them GC-eligible (assuming no other references hold the same String)
+     * instead of letting them live for the rest of the JVM in long-running
+     * embedded scenarios (CWE-316: Cleartext Storage of Sensitive Information
+     * in Memory).
+     * <p>
+     * <b>Caller-invoked.</b> This method must NOT be auto-invoked from
+     * {@link #execute()} — doing so (as the original CWE-316 fix did) breaks any
+     * caller that calls {@code execute()} more than once on the same scope, since
+     * the second call reads {@code "*****"} instead of the real credential.
+     * Callers that reuse a scope (e.g. status-before / update / status-after
+     * patterns) must call this method <b>after the final</b> {@code execute()},
+     * never between executions. Callers that use a scope only once (such as the
+     * CLI in {@code Main}) typically wrap the {@code execute()} call in a
+     * try-finally and call this in the {@code finally}.
      */
-    void clearCredentialArguments() {
+    public void clearCredentialArguments() {
         if (argumentValues.isEmpty()) {
             return;
         }

--- a/liquibase-standard/src/main/java/liquibase/integration/commandline/Main.java
+++ b/liquibase-standard/src/main/java/liquibase/integration/commandline/Main.java
@@ -1478,7 +1478,7 @@ public class Main {
                 // otherwise, we will print the output
                 //
                 Writer outputWriter = getOutputWriter();
-                CommandResults commandResults = snapshotCommand.execute();
+                CommandResults commandResults = executeAndClearCredentials(snapshotCommand);
                 String result = InternalSnapshotCommandStep.printSnapshot(snapshotCommand, commandResults);
                 outputWriter.write(result);
                 outputWriter.flush();
@@ -1521,7 +1521,7 @@ public class Main {
                         .addArgumentValue(DropAllCommandStep.CATALOG_AND_SCHEMAS_ARG, InternalSnapshotCommandStep.parseSchemas(database, getSchemaParams(database)))
                         .addArgumentValue(GenerateChangelogCommandStep.CHANGELOG_FILE_ARG, changeLogFile);
 
-                dropAllCommand.execute();
+                executeAndClearCredentials(dropAllCommand);
                 return;
             } else if (COMMANDS.STATUS.equalsIgnoreCase(command)) {
                 boolean runVerbose = false;
@@ -1558,7 +1558,7 @@ public class Main {
                         .addArgumentValue(CalculateChecksumCommandStep.CHANGESET_IDENTIFIER_ARG, getCommandParam(OPTIONS.CHANGE_SET_IDENTIFIER, null))
                         .addArgumentValue(CalculateChecksumCommandStep.CHANGELOG_FILE_ARG, this.changeLogFile);
 
-                calculateChecksumCommand.execute();
+                executeAndClearCredentials(calculateChecksumCommand);
                 return;
             } else if (COMMANDS.DB_DOC.equalsIgnoreCase(command)) {
                 if (commandParams.isEmpty()) {
@@ -1704,7 +1704,7 @@ public class Main {
                     historyCommand.addArgumentValue(HistoryCommandStep.FORMAT_ARG, HistoryFormat.valueOf(format));
                     historyCommand.setOutput(getOutputStream());
 
-                    historyCommand.execute();
+                    executeAndClearCredentials(historyCommand);
                 } else {
                     throw new CommandLineParsingException(
                             String.format(coreBundle.getString("command.unknown"), command));
@@ -1782,13 +1782,13 @@ public class Main {
         if (outputWriter != null) {
             commandScope.setOutput(new WriterOutputStream(outputWriter, GlobalConfiguration.OUTPUT_FILE_ENCODING.getCurrentValue()));
         }
-        commandScope.execute();
+        executeAndClearCredentials(commandScope);
     }
 
     private void runReleaseLocksCommand() throws CommandExecutionException {
         CommandScope commandScope = new CommandScope(ReleaseLocksCommandStep.COMMAND_NAME[0]);
         this.setDatabaseArgumentsToCommand(commandScope);
-        commandScope.execute();
+        executeAndClearCredentials(commandScope);
     }
 
     private void runGenerateChangelogCommandStep() throws LiquibaseException, IOException, CommandLineParsingException {
@@ -1827,7 +1827,7 @@ public class Main {
         this.setDatabaseArgumentsToCommand(diffChangelogCommand);
         this.setReferenceDatabaseArgumentsToCommand(diffChangelogCommand);
 
-        diffChangelogCommand.execute();
+        executeAndClearCredentials(diffChangelogCommand);
     }
 
     private void runDiffCommandStep() throws CommandLineParsingException, CommandExecutionException, IOException {
@@ -1839,7 +1839,7 @@ public class Main {
         this.setDatabaseArgumentsToCommand(diffCommand);
         this.setReferenceDatabaseArgumentsToCommand(diffCommand);
 
-        diffCommand.execute();
+        executeAndClearCredentials(diffCommand);
     }
 
     private void runUpdateCommandStep() throws CommandLineParsingException, CommandExecutionException, IOException {
@@ -1850,7 +1850,7 @@ public class Main {
         updateCommand.addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_CLASS_ARG, changeExecListenerClass);
         updateCommand.addArgumentValue(ChangeExecListenerCommandStep.CHANGE_EXEC_LISTENER_PROPERTIES_FILE_ARG, changeExecListenerPropertiesFile);
         setDatabaseArgumentsToCommand(updateCommand);
-        updateCommand.execute();
+        executeAndClearCredentials(updateCommand);
     }
 
     private void runRollbackOneChangeSetCommandStep() throws CommandExecutionException, CommandLineParsingException {
@@ -1865,7 +1865,7 @@ public class Main {
                 .addArgumentValue("force", getCommandParam(OPTIONS.FORCE, null));
         String internalCommand = "rollbackOneChangeset";
         Scope.getCurrentScope().addMdcValue(MdcKey.LIQUIBASE_INTERNAL_COMMAND, internalCommand);
-        rollbackOneChangeSet.execute();
+        executeAndClearCredentials(rollbackOneChangeSet);
     }
 
     private void runRollbackOneChangeSetSqlCommandStep() throws CommandExecutionException, CommandLineParsingException, IOException {
@@ -1881,7 +1881,7 @@ public class Main {
         String internalCommand = "rollbackOneChangeset";
         Scope.getCurrentScope().addMdcValue(MdcKey.LIQUIBASE_INTERNAL_COMMAND, internalCommand);
         rollbackOneChangeSet.setOutput(new WriterOutputStream(getOutputWriter(), GlobalConfiguration.OUTPUT_FILE_ENCODING.getCurrentValue()));
-        rollbackOneChangeSet.execute();
+        executeAndClearCredentials(rollbackOneChangeSet);
     }
 
     private void runRollbackOneUpdateCommandStep() throws CommandExecutionException, CommandLineParsingException {
@@ -1894,7 +1894,7 @@ public class Main {
                 .addArgumentValue("force", getCommandParam(OPTIONS.FORCE, null));
         String internalCommand = "rollbackOneUpdate";
         Scope.getCurrentScope().addMdcValue(MdcKey.LIQUIBASE_INTERNAL_COMMAND, internalCommand);
-        rollbackOneUpdate.execute();
+        executeAndClearCredentials(rollbackOneUpdate);
     }
 
     private void runRollbackOneUpdateSqlCommandStep() throws CommandExecutionException, CommandLineParsingException, IOException {
@@ -1908,7 +1908,7 @@ public class Main {
         String internalCommand = "rollbackOneUpdate";
         Scope.getCurrentScope().addMdcValue(MdcKey.LIQUIBASE_INTERNAL_COMMAND, internalCommand);
         rollbackOneUpdate.setOutput(new WriterOutputStream(getOutputWriter(), GlobalConfiguration.OUTPUT_FILE_ENCODING.getCurrentValue()));
-        rollbackOneUpdate.execute();
+        executeAndClearCredentials(rollbackOneUpdate);
     }
 
     /**
@@ -1989,7 +1989,7 @@ public class Main {
         executeSql.addArgumentValue(ExecuteSqlCommandStep.SQLFILE_ARG, getCommandParam("sqlFile", null));
         executeSql.addArgumentValue(ExecuteSqlCommandStep.DELIMITER_ARG, getCommandParam("delimiter", ";"));
         setDatabaseArgumentsToCommand(executeSql);
-        executeSql.execute();
+        executeAndClearCredentials(executeSql);
     }
 
     /**
@@ -2067,6 +2067,29 @@ public class Main {
     }
 
     /**
+     * Execute the given {@link CommandScope} and unconditionally call
+     * {@link CommandScope#clearCredentialArguments()} on it afterwards (CWE-316:
+     * Cleartext Storage of Sensitive Information in Memory). Used in place of a
+     * bare {@code commandScope.execute()} at every CLI command-runner site so
+     * that single-use scopes drop their credential references before this method
+     * returns, regardless of command outcome.
+     * <p>
+     * The clearing was originally inlined into {@link CommandScope#execute()}'s
+     * outer {@code finally}, but that broke any caller that calls {@code execute()}
+     * more than once on the same scope (the second call sees {@code "*****"} and
+     * the JDBC driver rejects). Moving the clearing to the CLI entry points
+     * preserves the CWE-316 intent here without imposing single-use semantics on
+     * programmatic / library callers.
+     */
+    private static CommandResults executeAndClearCredentials(CommandScope commandScope) throws CommandExecutionException {
+        try {
+            return commandScope.execute();
+        } finally {
+            commandScope.clearCredentialArguments();
+        }
+    }
+
+    /**
      * Return the first "parameter" from the command line that does NOT have the form of parameter=value. A parameter
      * is a command line argument that follows the main action (e.g. update/rollback/...). Example:
      * For the main action "updateToTagSQL &lt;tag&gt;", &lt;tag&gt; would be the command argument.
@@ -2137,7 +2160,7 @@ public class Main {
             commandScope.addArgumentValue("exception", exception);
             commandScope.addArgumentValue("listener", defaultChangeExecListener);
             commandScope.addArgumentValue("rollbackOnError", rollbackOnError);
-            commandScope.execute();
+            executeAndClearCredentials(commandScope);
         } catch (IllegalArgumentException ignoredCommandNotFound) {
             throw exception;
         }

--- a/liquibase-standard/src/main/java/liquibase/integration/commandline/Main.java
+++ b/liquibase-standard/src/main/java/liquibase/integration/commandline/Main.java
@@ -1809,6 +1809,7 @@ public class Main {
 
         this.setDatabaseArgumentsToCommand(generateChangelogCommand);
         this.setPreCompareArgumentsToCommand(generateChangelogCommand);
+        executeAndClearCredentials(generateChangelogCommand);
     }
 
     private void runDiffChangelogCommandStep() throws CommandExecutionException, CommandLineParsingException, IOException {

--- a/liquibase-standard/src/test/groovy/liquibase/command/CommandScopeTest.groovy
+++ b/liquibase-standard/src/test/groovy/liquibase/command/CommandScopeTest.groovy
@@ -196,13 +196,19 @@ class CommandScopeTest extends Specification {
         Locale.setDefault(savedDefault)
     }
 
-    def "execute invokes credential clearing in its outer finally on the success path"() {
-        // CWE-316 wiring check: confirms the outer finally added to execute()
-        // actually calls clearCredentialArguments(). Pre-supplies the 'requiredArg'
-        // value so this test is order-independent w.r.t. the 'execute with failing
-        // argument validation' spec that registers requiredArg as required on
-        // 'mock' (Spock test order in a single Specification is declaration-order,
-        // but the side effect of CommandBuilder is global to the JVM).
+    def "execute does NOT clear credential arguments automatically — opt-in semantics"() {
+        // Regression: PR #7741 originally clamped clearCredentialArguments() into
+        // execute()'s outer finally to satisfy CWE-316. That broke any caller that
+        // re-uses a CommandScope (see PostgreSQLIntegrationTest.testStatusRunDuringUpdate
+        // which executes 3× on the same scope: first call succeeded; second/third saw
+        // "*****" in argumentValues["password"] and got "FATAL: password authentication
+        // failed for user 'lbuser'" from the JDBC driver). The fix moved the clearing
+        // out of execute() and made it caller-invoked via the (now public)
+        // clearCredentialArguments() method. This spec pins that execute() alone does
+        // NOT touch credential argument values, so re-use cases continue to work.
+        // Pre-supplies 'requiredArg' so the test is order-independent w.r.t. the
+        // 'execute with failing argument validation' spec above which registers
+        // requiredArg as required on 'mock'.
         given:
         def scope = new CommandScope("mock")
         scope.addArgumentValue("requiredArg", "anything")
@@ -212,7 +218,7 @@ class CommandScopeTest extends Specification {
         MockCommandStep.logic = new MockCommandStep() {
             @Override
             void run(CommandResultsBuilder resultsBuilder) throws Exception {
-                // no-op pipeline body — we are testing the finally-block wiring, not the step.
+                // no-op pipeline body — we are pinning the absence of auto-clearing, not the step.
             }
         }
 
@@ -220,13 +226,51 @@ class CommandScopeTest extends Specification {
         scope.execute()
 
         then:
-        scope.@argumentValues["password"]    == "*****"
+        scope.@argumentValues["password"]    == "supersecret-pw-12345"
         scope.@argumentValues["requiredArg"] == "anything"
         scope.@argumentValues["username"]    == "regular-user"
     }
 
-    def "execute invokes credential clearing in its outer finally even when the pipeline throws"() {
-        // CWE-316 wiring check: the outer finally must run on the failure path too.
+    def "execute called twice on the same scope reads original credentials both times — regression for PostgreSQLIntegrationTest.testStatusRunDuringUpdate"() {
+        // The canary scenario that surfaced the auto-clear regression. A caller
+        // creates a CommandScope, executes it (e.g., 'status'), then re-executes
+        // the SAME scope (e.g., 'update', or 'status' again after a concurrent
+        // update). With the old auto-clear in execute()'s finally, the second
+        // execute() would see argumentValues["password"] == "*****" and the JDBC
+        // driver would reject the connection with "password authentication failed".
+        // Post-fix: both executions see the original credential.
+        given:
+        def scope = new CommandScope("mock")
+        scope.addArgumentValue("requiredArg", "anything")
+        scope.addArgumentValue("password",    "supersecret-pw-12345")
+
+        def passwordsObservedByPipeline = []
+        MockCommandStep.logic = new MockCommandStep() {
+            @Override
+            void run(CommandResultsBuilder resultsBuilder) throws Exception {
+                // Capture what the pipeline would have read at execute() time.
+                passwordsObservedByPipeline << resultsBuilder.getCommandScope().@argumentValues["password"]
+            }
+        }
+
+        when:
+        scope.execute()
+        scope.execute()
+        scope.execute()
+
+        then:
+        // All three executions saw the same original credential — none was redacted
+        // between calls. This is the contract that re-use callers depend on.
+        passwordsObservedByPipeline == ["supersecret-pw-12345", "supersecret-pw-12345", "supersecret-pw-12345"]
+        scope.@argumentValues["password"] == "supersecret-pw-12345"
+    }
+
+    def "execute does NOT clear credentials even when the pipeline throws — opt-in semantics also apply on the failure path"() {
+        // Counterpart to the success-path spec above: under the new opt-in semantics,
+        // execute() must NOT auto-clear on exception either. If a caller needs the
+        // CWE-316 protection on the failure path, they wrap the execute() call in
+        // try-finally and call clearCredentialArguments() themselves (see
+        // Main.executeAndClearCredentials()).
         given:
         def scope = new CommandScope("mock")
         scope.addArgumentValue("requiredArg", "anything")
@@ -244,6 +288,37 @@ class CommandScopeTest extends Specification {
 
         then:
         thrown(liquibase.exception.CommandExecutionException)
+        scope.@argumentValues["password"] == "supersecret-pw-12345"
+    }
+
+    def "caller-invoked try-finally pattern around execute clears credentials post-execution (CWE-316 wiring documented for CLI integrators)"() {
+        // The pattern recommended for CLI / single-use callers: wrap execute() in
+        // try-finally and call clearCredentialArguments() in the finally. This spec
+        // documents the contract: the pipeline sees the original credential during
+        // execute(), and the credential is wiped immediately after the call returns.
+        // Main.executeAndClearCredentials() is the canonical implementation.
+        given:
+        def scope = new CommandScope("mock")
+        scope.addArgumentValue("requiredArg", "anything")
+        scope.addArgumentValue("password",    "supersecret-pw-12345")
+
+        def passwordSeenByPipeline = null
+        MockCommandStep.logic = new MockCommandStep() {
+            @Override
+            void run(CommandResultsBuilder resultsBuilder) throws Exception {
+                passwordSeenByPipeline = resultsBuilder.getCommandScope().@argumentValues["password"]
+            }
+        }
+
+        when:
+        try {
+            scope.execute()
+        } finally {
+            scope.clearCredentialArguments()
+        }
+
+        then:
+        passwordSeenByPipeline == "supersecret-pw-12345"
         scope.@argumentValues["password"] == "*****"
     }
 

